test/rand_tangent.jl

```diff
@@ -0,0 +1,85 @@
+# Test struct for `rand_tangent`.
+struct Foo
+   a::Float64
+   b::Int
+   c::Any
+end
+
+@testset "generate_tangent" begin
+    rng = MersenneTwister(123456)
+
+    foreach([
+
+        # Things without sensible tangents.
```

Comment:
I think it would be more. idiomatic if we used a `for` loop (rather than `foreach(...) 

I also think the these tests would be clearer if we turned these comments into testsets e.g. something along the lines of
```julia
@testset "rand_tangent" begin
    rng = MersenneTwister(123456)
   
    @testset  "No tangent: $(typeof(x))" for x in ("hi", 'a', ...)
        @test rand_tangent(rng, x) isa DoesNotExist()
        @test rand_tangent(x) isa DoesNotExist()
        @test x + rand_tangent(rng, x) isa typeof(x)
    end

    @testset "Numbers: $(typeof(x))" for x in (5.0, 5.0 + 0.4im)
        @test rand_tangent(rng, x) isa typeof(x)
        @test rand_tangent(x) isa typeof(x)
        @test x + rand_tangent(rng, x) isa typeof(x)
    end  
    
    @testset "etc..."
end
```
Obviously something like this is more verbose, but I think it would be clearer

Reply:
Hmm I'd rather avoid a lot of copy + pasted code. Would something like:
```julia

    @testset "Primal: $(typeof(x)), Tangent: $T_tangent" for (x, T_tangent) in [

        # Things without sensible tangents.
        ("hi", DoesNotExist),
        ('a', DoesNotExist),
        (:a, DoesNotExist),
        (true, DoesNotExist),
        (4, DoesNotExist),
        ...
    ]
        @test rand_tangent(rng, x) isa T_tangent
        @test rand_tangent(x) isa T_tangent
        @test x + rand_tangent(rng, x) isa typeof(x)
    end

```
be acceptable?

Reply:
yes, i think that's an improvement as it'll give us a more informative message if something fails. thanks